sources/file.go

```diff
@@ -0,0 +1,219 @@
+package sources
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/h2non/filetype"
+	"github.com/mholt/archives"
+
+	"github.com/zricethezav/gitleaks/v8/logging"
+)
+
+const defaultBufferSize = 100 * 1_000 // 100kb
+const InnerPathSeparator = "!"
+
+type seekReaderAt interface {
+	io.ReaderAt
+	io.Seeker
+}
+
+// File implements Source for scanning a reader with a path
+type File struct {
+	// Content provides a reader to the file's content
+	Content io.Reader
+	// Path is the resolved real path of the file
+	Path string
+	// Symlink represents a symlink to the file if that's how it was discovered
+	Symlink string
+	// Buffer is used for reading the content in chunks
+	Buffer []byte
+	// outerPaths is the list of container paths (e.g. archives) that lead to
+	// this file
+	outerPaths []string
+}
+
+// Fragments yields fragments for the this source
+func (s *File) Fragments(yield FragmentsFunc) error {
+	ctx := context.Background()
```

Comment:
IMO we should try to support contexts for new APIs.

See: https://github.com/gitleaks/gitleaks/pull/1841#discussion_r2071559066 and #1849.

Reply:
+1 

@zricethezav want me to go through and change `Fragments(FragmentsFunc)` -> `Fragments(context.Context, FragmentsFunc)` ?